src/theory/sep/theory_sep.cpp

```diff
@@ -1341,7 +1342,7 @@ Node TheorySep::mkUnion( TypeNode tn, std::vector< Node >& locs ) {
     for( unsigned i=0; i<locs.size(); i++ ){
       Node s = locs[i];
       Assert(!s.isNull());
-      s = NodeManager::currentNM()->mkNode( kind::SINGLETON, s );
+      s = NodeManager::currentNM()->mkSingleton(tn, s);
```

Comment:
Here I used tn which is an argument to `mkUnion` and the type of the elements. This function returns either an empty set, or a union of sets starting from a singleton. Perhaps it is much simpler to start from an empty set (which stores the type).